server/events.go

```diff
@@ -102,32 +162,356 @@ func (s *Server) internalSendLoop() {
+var orphanServerDuration = 5 * connHBInterval
+var checkRemoteServerInterval = 3 * connHBInterval
+
+// Check for orphan servers who may have gone away without notification.
+func (s *Server) checkRemoteServers() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.eventsEnabled() {
+		return
+	}
+	now := time.Now()
+	for sid, su := range s.sys.servers {
+		if now.Sub(su.ltime) > orphanServerDuration {
+			s.Debugf("Detected orphan remote server: %q", sid)
+			// Simulate it going away.
+			s.processRemoteServerShutdown(sid)
+			delete(s.sys.servers, sid)
+		}
+	}
+	s.sys.sweeper.Reset(checkRemoteServerInterval)
+}
+
+// Start a ticker that will fire periodically and check for orphaned servers.
+func (s *Server) startRemoteServerSweepTimer() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.eventsEnabled() {
+		return
+	}
+	s.sys.sweeper = time.AfterFunc(checkRemoteServerInterval, s.checkRemoteServers)
+}
+
+// This will setup our system wide tracking subs.
+// For now we will setup one wildcard subscription to
+// monitor all accounts for changes in number of connections.
+// We can make this on a per account tracking basis if needed.
+// Tradeoff is subscription and interest graph events vs connect and
+// disconnect events, etc.
+func (s *Server) initEventTracking() {
+	if !s.eventsEnabled() {
+		return
+	}
+	subject := fmt.Sprintf(accConnsEventSubj, "*")
+	if _, err := s.sysSubscribe(subject, s.remoteConnsUpdate); err != nil {
+		s.Errorf("Error setting up internal tracking: %v", err)
+	}
+	// This will be for responses for account info that we send out.
+	subject = fmt.Sprintf(connsRespSubj, s.info.ID)
+	if _, err := s.sysSubscribe(subject, s.remoteConnsUpdate); err != nil {
+		s.Errorf("Error setting up internal tracking: %v", err)
+	}
+	// Listen for broad requests to respond with account info.
+	subject = fmt.Sprintf(accConnsReqSubj, "*")
+	if _, err := s.sysSubscribe(subject, s.connsRequest); err != nil {
+		s.Errorf("Error setting up internal tracking: %v", err)
+	}
+	// Listen for all server shutdowns.
+	subject = fmt.Sprintf(shutdownEventSubj, "*")
+	if _, err := s.sysSubscribe(subject, s.remoteServerShutdown); err != nil {
+		s.Errorf("Error setting up internal tracking: %v", err)
+	}
+}
+
+// processRemoteServerShutdown will update any affected accounts.
+// Will upidate the remote count for clients.
+// Lock assume held.
+func (s *Server) processRemoteServerShutdown(sid string) {
+	for _, a := range s.accounts {
+		a.mu.Lock()
+		prev := a.strack[sid]
+		delete(a.strack, sid)
+		a.nrclients -= prev
+		a.mu.Unlock()
+	}
+}
+
+// serverShutdownEvent is called when we get an event from another server shutting down.
+func (s *Server) remoteServerShutdown(sub *subscription, subject, reply string, msg []byte) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.eventsEnabled() {
+		return
+	}
+	toks := strings.Split(subject, tsep)
+	if len(toks) != shutdownEventTokens {
+		s.Debugf("Received remote server shutdown on bad subject %q", subject)
```

Comment:
Can make '<'

Reply:
Point is that it should not be possible since this is a message callback linked to subject we know has the number of tokens. But again, if you want defensive code I understand, but we should not continue after Debugf...

Reply:
in other words, there should probably be a return here...

Reply:
yep